core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/ldap/ReadOnlyLDAPUserStoreManager.java

```diff
@@ -2639,6 +2650,462 @@ public int getTenantId() throws UserStoreException {
+                            Control.CRITICAL)});
+                } while ((cookie != null) && (cookie.length != 0));
+            }
+        } catch (PartialResultException e) {
+            // Can be due to referrals in AD. So just ignore error
+            String errorMessage = "Error occurred while searching for user(s) for filter: " + searchFilter;
+            if (isIgnorePartialResultException()) {
+                if (log.isDebugEnabled()) {
+                    log.debug(errorMessage, e);
+                }
+            } else {
+                throw new UserStoreException(errorMessage, e);
+            }
+        } catch (NamingException e) {
+            String errorMessage = "Error occurred while searching for user(s) for filter: " + searchFilter;
+            if (log.isDebugEnabled()) {
+                log.debug(errorMessage, e);
+            }
+            throw new UserStoreException(errorMessage, e);
+        } catch (IOException e) {
+            e.printStackTrace();
+        } finally {
+            // Close ldap context, directory context, close naming enumeration and free up the resources.
+            JNDIUtil.closeContext(dirContext);
+            JNDIUtil.closeContext(ldapContext);
+            JNDIUtil.closeNamingEnumeration(answer);
+        }
+        result.setUsers(users.toArray(new String[0]));
+        return result;
+    }
+
+    private static byte[] parseControls(Control[] controls) {
+
+        byte[] cookie = null;
+        // Handle the paged results control response
+        if (controls != null) {
+            for (int i = 0; i < controls.length; i++) {
+                if (controls[i] instanceof PagedResultsResponseControl) {
+                    PagedResultsResponseControl prrc = (PagedResultsResponseControl) controls[i];
+                    cookie = prrc.getCookie();
+                }
+            }
+        }
+        return cookie;
+    }
+
+    private List<ExpressionCondition> getExpressionConditions(Condition condition) {
+
+        List<ExpressionCondition> expressionConditions = new ArrayList<>();
+        getExpressionConditionsAsList(condition, expressionConditions);
+        return expressionConditions;
+    }
+
+    private void getExpressionConditionsAsList(Condition condition, List<ExpressionCondition> expressionConditions) {
+
+        if (condition instanceof ExpressionCondition) {
+            expressionConditions.add((ExpressionCondition) condition);
+        } else if (condition instanceof OperationalCondition) {
+            Condition leftCondition = ((OperationalCondition) condition).getLeftCondition();
+            getExpressionConditionsAsList(leftCondition, expressionConditions);
+            Condition rightCondition = ((OperationalCondition) condition).getRightCondition();
+            getExpressionConditionsAsList(rightCondition, expressionConditions);
+        }
+    }
+
+    private List<String> getUserListFromGroupFilteringResult(NamingEnumeration<SearchResult> answer,
```

Comment:
This getUserListFromGroupFilteringResult() method is very long, please break this into several but meaningful methods, ideally each method should have only 20/30 LOC.

Reply:
Fixed.